// ngx-uploadx/esm2015/backoff_retry.js
/**
 *  Exponential Backoff Retries
 */
export class BackoffRetry {
    /**
     * @param min  Initial retry delay
     * @param max  Max retry delay
     * @param k    Increase factor
     */
    constructor(min = 200, max = min * 300, k = 2) {
        this.min = min;
        this.max = max;
        this.k = k;
        this.code = -1;
        this.retryAttempts = 1;
        this.delay = this.min;
    }
    /**
     * Delay Retry
     * @param code
     * @returns retryAttempts
     */
    wait(code) {
        return new Promise(resolve => {
            if (code === this.code) {
                this.retryAttempts++;
                this.delay = Math.min(this.delay * this.k, this.max);
            }
            else {
                this.delay = this.min;
                this.retryAttempts = 1;
            }
            this.code = code;
            setTimeout(() => resolve(this.retryAttempts), this.delay + Math.floor(Math.random() * this.min));
        });
    }
    /**
     * Reset Retry
     */
    reset() {
        this.delay = this.min;
        this.retryAttempts = 1;
        this.code = -1;
    }
}
