// ngx-uploadx/esm2015/uploadx.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Renderer2, Input, Output, EventEmitter } from '@angular/core';
import { UploadxOptions } from './interfaces';
import { UploadxService } from './uploadx.service';
let UploadxDirective = class UploadxDirective {
    constructor(elementRef, renderer, uploadService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.uploadService = uploadService;
        this.uploadxState = new EventEmitter();
        this.fileListener = () => {
            if (this.elementRef.nativeElement.files) {
                this.uploadService.handleFileList(this.elementRef.nativeElement.files);
            }
        };
    }
    set uploadxAction(ctrlEvent) {
        if (ctrlEvent && this.uploadService) {
            this.uploadService.control(ctrlEvent);
        }
    }
    ngOnInit() {
        if (this.uploadx) {
            if (this.uploadx.allowedTypes) {
                this.renderer.setAttribute(this.elementRef.nativeElement, 'accept', this.uploadx.allowedTypes);
            }
            this.uploadService.init(this.uploadx);
        }
        this.uploadxState.emit(this.uploadService.events);
        this.listenerFn = this.renderer.listen(this.elementRef.nativeElement, 'change', this.fileListener);
    }
    ngOnDestroy() {
        if (this.listenerFn) {
            this.listenerFn();
        }
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], UploadxDirective.prototype, "uploadxState", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", UploadxOptions)
], UploadxDirective.prototype, "uploadx", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], UploadxDirective.prototype, "uploadxAction", null);
UploadxDirective = tslib_1.__decorate([
    Directive({
        selector: '[uploadx]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef,
        Renderer2,
        UploadxService])
], UploadxDirective);
export { UploadxDirective };
