// ngx-uploadx/esm5/backoff_retry.js
/**
 *  Exponential Backoff Retries
 */
var BackoffRetry = /** @class */ (function () {
    /**
     * @param min  Initial retry delay
     * @param max  Max retry delay
     * @param k    Increase factor
     */
    function BackoffRetry(min, max, k) {
        if (min === void 0) { min = 200; }
        if (max === void 0) { max = min * 300; }
        if (k === void 0) { k = 2; }
        this.min = min;
        this.max = max;
        this.k = k;
        this.code = -1;
        this.retryAttempts = 1;
        this.delay = this.min;
    }
    /**
     * Delay Retry
     * @param code
     * @returns retryAttempts
     */
    BackoffRetry.prototype.wait = function (code) {
        var _this = this;
        return new Promise(function (resolve) {
            if (code === _this.code) {
                _this.retryAttempts++;
                _this.delay = Math.min(_this.delay * _this.k, _this.max);
            }
            else {
                _this.delay = _this.min;
                _this.retryAttempts = 1;
            }
            _this.code = code;
            setTimeout(function () { return resolve(_this.retryAttempts); }, _this.delay + Math.floor(Math.random() * _this.min));
        });
    };
    /**
     * Reset Retry
     */
    BackoffRetry.prototype.reset = function () {
        this.delay = this.min;
        this.retryAttempts = 1;
        this.code = -1;
    };
    return BackoffRetry;
}());
export { BackoffRetry };
